File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/CamGirl.java
package org.firstinspires.ftc.teamcode;

import static org.firstinspires.ftc.robotcore.external.BlocksOpModeCompanion.telemetry;
import static org.firstinspires.ftc.teamcode.RobotFuncs.batteryVoltageSensor;
import static org.firstinspires.ftc.teamcode.RobotFuncs.dashboard;
import static org.firstinspires.ftc.teamcode.RobotFuncs.hardwareMap;

import com.acmerobotics.dashboard.telemetry.TelemetryPacket;

import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName;
import org.firstinspires.ftc.teamcode.trajectorysequence.TrajectorySequence;
import org.openftc.apriltag.AprilTagDetection;
import org.openftc.easyopencv.OpenCvCamera;
import org.openftc.easyopencv.OpenCvCameraFactory;
import org.openftc.easyopencv.OpenCvCameraRotation;

import java.util.ArrayList;

public class CamGirl implements Runnable{
    public int LAST_ID = 0;
    public boolean shouldClose = false;
    OpenCvCamera webcam;

    public CamGirl(OpenCvCamera cm) {
        this.webcam = cm;
    }

    boolean OPENED = false;
    @Override
    public void run() {
        AprilTagDetectionPipeline pipeline;
        double TAGSIZE = 4.5 / 100;
        double FX = 878.272;
        double FY = 878.272;
        double CX = 320;
        double CY = 240;

        TelemetryPacket p;

        p = new TelemetryPacket();
        p.addLine("MKP");
        dashboard.sendTelemetryPacket(p);

        pipeline = new AprilTagDetectionPipeline(TAGSIZE, FX, FY, CX, CY);
        p = new TelemetryPacket();
        p.addLine("SETP");
        dashboard.sendTelemetryPacket(p);
        webcam.setPipeline(pipeline);

        p = new TelemetryPacket();
        p.addLine("OPEN");
        dashboard.sendTelemetryPacket(p);
        webcam.openCameraDeviceAsync(new OpenCvCamera.AsyncCameraOpenListener() {
            @Override
            public void onOpened() {
                webcam.startStreaming(640, 480, OpenCvCameraRotation.SIDEWAYS_LEFT);
                dashboard.startCameraStream(webcam, 15);
                OPENED = true;
                TelemetryPacket pack = new TelemetryPacket();
                pack.addLine("OPENED");
                dashboard.sendTelemetryPacket(pack);
            }

            @Override
            public void onError(int errorCode) {
                TelemetryPacket pack = new TelemetryPacket();
                pack.put("Erro", errorCode);
                dashboard.sendTelemetryPacket(pack);
            }
        });

        while (shouldClose) {
            if (OPENED) {
                if (LAST_ID == 0) {
                    TelemetryPacket packet = new TelemetryPacket();
                    packet.put("Waiting on cam open", 0);
                    dashboard.sendTelemetryPacket(packet);
                }
                ArrayList<AprilTagDetection> cd = pipeline.getLatestDetections();
                if (cd.size() > 0) {
                    LAST_ID = cd.get(0).id;
                    TelemetryPacket packet = new TelemetryPacket();
                    packet.put("LID", LAST_ID);
                    dashboard.sendTelemetryPacket(packet);
                }
            }
            try {
                Thread.sleep(5);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        p = new TelemetryPacket();
        p.addLine("Close");
        dashboard.sendTelemetryPacket(p);

        if (shouldClose && OPENED) {
            webcam.closeCameraDeviceAsync(() -> {
            });
        }
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/mk3/camera/CamGirl.kt
package org.firstinspires.ftc.teamcode.mk3.camera

import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import org.firstinspires.ftc.robotcore.external.hardware.camera.WebcamName
import org.firstinspires.ftc.robotcore.external.hardware.camera.controls.ExposureControl
import org.openftc.easyopencv.*
import java.lang.Thread.sleep
import java.util.concurrent.TimeUnit

/// Thanks to Brainstormz for generously "gifting" this code to us

class CamGirl(lom: LinearOpMode,
              name: String,
              orientation: OpenCvCameraRotation,
              resX: Int,
              resY: Int,
              pipeline: OpenCvPipeline,
              streaming: Boolean,
              waitForOpen: Boolean,
              gain: Int,
              exposure: Int) {

    constructor(lom: LinearOpMode,
                name: String,
                orientation: OpenCvCameraRotation,
                resX: Int,
                resY: Int,
                pipeline: OpenCvPipeline
    ) : this(lom, name, orientation, resX, resY, pipeline, false, false)

    constructor(lom: LinearOpMode,
                name: String,
                orientation: OpenCvCameraRotation,
                resX: Int,
                resY: Int,
                pipeline: OpenCvPipeline,
                streaming: Boolean,
                waitForOpen: Boolean,
    ) : this(lom, name, orientation, resX, resY, pipeline, streaming, waitForOpen, 100, 0)

    var camera: OpenCvWebcam
    var ecode: Int = 0
    var opened: Boolean = false
    var expos: Int = 10
    private var dashboardStreaming = false

    init {
        val cameraMonitorViewId: Int = lom.hardwareMap.appContext.resources.getIdentifier("cameraMonitorViewId", "id", lom.hardwareMap.appContext.packageName)
        val webcamName: WebcamName = lom.hardwareMap.get(WebcamName::class.java, name)
        camera = OpenCvCameraFactory.getInstance().createWebcam(webcamName, cameraMonitorViewId)
        camera.setPipeline(pipeline)

        expos = exposure

        dashboardStreaming = streaming
        val cameraListener = object : OpenCvCamera.AsyncCameraOpenListener {
            override fun onOpened() {
                if (exposure != 0) {
                    camera.exposureControl.mode = ExposureControl.Mode.Manual
                    camera.exposureControl.setExposure(expos.toLong(), TimeUnit.MILLISECONDS)
                }
                camera.gainControl.gain = gain

                camera.startStreaming(resX, resY, orientation)
                if (streaming) {
                    FtcDashboard.getInstance().startCameraStream(camera, 30.0)
                }
                opened = true
            }

            override fun onError(errorCode: Int) {
                ecode = errorCode
            }
        }

        camera.openCameraDeviceAsync(cameraListener)
        while (waitForOpen && !opened && !lom.isStopRequested) {
            val tp = TelemetryPacket()
            tp.addLine("Currently waiting on cam open")
            FtcDashboard.getInstance().sendTelemetryPacket(tp)
            lom.telemetry.addLine("Waiting on cam open")
            lom.telemetry.update()
            sleep(5)
        }
        lom.telemetry.addLine("Cam opened")
        lom.telemetry.update()

    }

    fun stop() {
        camera.stopStreaming()
        if (dashboardStreaming) {
            FtcDashboard.getInstance().stopCameraStream();
        }
    }
}